Add check_mode tests for shell/command/script and make skipped=True/changed=True mutually exclusive

s-hertel opened the pull request:
##### SUMMARY
Follow up to  https://github.com/ansible/ansible/pull/76353

##### ISSUE TYPE
- Test Pull Request

s-hertel committed:
Add additional tests for check_mode with shell/command/script

bcoca commented on test/integration/targets/script/tasks/main.yml:
```diff
@@ -177,6 +177,7 @@
   assert:
     that:
       - _check_mode_test is changed
+      - _check_mode_test is skipped
```
this should not pass, only changed or skipped should be valid, not both

s-hertel committed:
update check_mode documentation: if a module/action is not skipped in check mode and accurately reflects whether a change is made to the remote support should be 'full'

s-hertel committed:
Make reporting skipped in check mode mutually exclusive with 'changed: True'

s-hertel committed:
Add missing documented attributes

s-hertel committed:
fix typo

s-hertel committed:
Fix tests to expect skipped=True and changed=True do not occur together

s-hertel committed:
Fix script check_mode support documentation

s-hertel committed:
Fix earlier changelog

s-hertel commented:
I went through the list of modules/actions with partial or no check mode support and found nothing other than command/shell (and script since #76353) returning 'changed: True' and 'skipped: True', so I fixed that and updated the tests. While fixing that for script, I realized that check mode should actually be partially supported, just like for command (via creates/removes). The large diff for script is just unindentation and looks better with `?w=1` added to the files url.

s-hertel committed:
document platforms attribute

bcoca commented on lib/ansible/plugins/action/script.py:
```diff
@@ -87,41 +87,47 @@ def run(self, tmp=None, task_vars=None):
             except AnsibleError as e:
                 raise AnsibleActionFail(to_native(e))
 
+            if self._play_context.check_mode:
```
use self._task instead (play_context is not updated per loop item)

s-hertel replied:
Fixed, good catch

s-hertel committed:
Use tasks's check mode since the value from PlayContext does not reflect loop items

bcoca commented on lib/ansible/plugins/action/script.py:
```diff
@@ -87,41 +87,47 @@ def run(self, tmp=None, task_vars=None):
+                # check mode is supported if 'creates' or 'removes' are provided
+                # the task has already been skipped if a change would not occur
+                if self._task.args.get('creates') or self._task.args.get('removes'):
+                    result['changed'] = True
```
`raise AnsibleActionDone(result=result)`

bcoca commented on lib/ansible/plugins/action/script.py:
```diff
@@ -87,41 +87,47 @@ def run(self, tmp=None, task_vars=None):
+                # but since the script may override 'changed', just skip instead of guessing.
+                else:
+                    result['changed'] = False
+                    raise AnsibleActionSkip('Check mode is not supported for this task.')
```
raise AnsibleActionSkip('Check mode is not supported for this task.', result=result)

s-hertel committed:
requested changes

Co-authored-by: Brian Coca <bcoca@users.noreply.github.com>

s-hertel commented:
rebuild_merge
